src/ucp/core/ucp_request.c

```diff
@@ -273,6 +274,101 @@ int ucp_request_pending_add(ucp_request_t *req, unsigned pending_flags)
+                                      &prog_id);
+}
+
+static void ucp_request_dt_invalidate(ucp_context_t *context,
+                                      uct_mem_h *uct_memh,
+                                      ucp_md_map_t *md_map_p,
+                                      ucp_request_t *req)
+{
+    uct_md_mem_dereg_params_t params = {
+        .field_mask = UCT_MD_MEM_DEREG_FIELD_MEMH |
+                      UCT_MD_MEM_DEREG_FIELD_FLAGS |
+                      UCT_MD_MEM_DEREG_FIELD_CALLBACK |
+                      UCT_MD_MEM_DEREG_FIELD_ARG,
+        .flags      = UCT_MD_MEM_DEREG_FLAG_INVALIDATE,
+        .cb         = ucp_request_dt_invalidate_cb,
+        .arg        = req
+    };
+    unsigned md_index;
+    unsigned memh_index;
+    ucs_status_t status;
+
+    memh_index = 0;
+    ucs_for_each_bit(md_index, *md_map_p) {
+        ucs_trace("invalidating memh[%d]=%p from md[%d]", memh_index,
+                  uct_memh[memh_index], md_index);
+        req->send.state.uct_comp.count++;
+        params.memh = uct_memh[memh_index];
+        status      = uct_md_mem_dereg_v2(context->tl_mds[md_index].md,
+                                          &params);
+        if (status != UCS_OK) {
+            ucs_warn("failed to dereg from md[%d]=%s: %s", md_index,
+                        context->tl_mds[md_index].rsc.md_name,
+                        ucs_status_string(status));
+            req->send.state.uct_comp.count--;
+        }
+        memh_index++;
+    }
+
+    *md_map_p = 0;
+}
+
+void ucp_request_send_err_dereg(ucp_request_t *req, ucs_status_t status)
+{
+    ucp_context_t *context = req->send.ep->worker->context;
+    ucp_md_map_t dereg_map = ucp_ep_config(req->send.ep)->key.rma_inv_md_map &
+                             req->send.state.dt.dt.contig.md_map;
+
+    ucs_assert(status != UCS_OK);
+    ucs_assert(ucp_ep_config(req->send.ep)->key.err_mode !=
+               UCP_ERR_HANDLING_MODE_NONE);
+    ucs_assert(ucp_ep_config(req->send.ep)->key.rma_inv_md_map);
+    ucs_assert(UCP_DT_IS_CONTIG(req->send.datatype));
+
+    req->send.state.uct_comp.count  = 1;
+    req->send.state.uct_comp.func   = ucp_request_err_uct_comp_cb;
+    req->send.state.uct_comp.status = status;
+
+    ucp_trace_req(req, "mem dereg buffer md_map 0x%"PRIx64, dereg_map);
+    ucp_mem_rereg_mds(context, dereg_map, NULL, 0, 0, NULL,
+                      UCS_MEMORY_TYPE_HOST, NULL,
+                      req->send.state.dt.dt.contig.memh,
+                      &req->send.state.dt.dt.contig.md_map);
+    ucp_trace_req(req, "mem invalidate buffer md_map 0x%"PRIx64,    
+                  req->send.state.dt.dt.contig.md_map);
+    ucp_request_dt_invalidate(context, req->send.state.dt.dt.contig.memh,
+                              &req->send.state.dt.dt.contig.md_map, req);
+    ucs_assert(req->send.state.dt.dt.contig.md_map == 0);
+    ucp_invoke_uct_completion(&req->send.state.uct_comp, status);
```

Comment:
on line 355 there is set counter to 1, and in case if any region is waiting invalidation then ucp_invoke_uct_completion just decrement counter without callback involving